packages/xod-client-browser/benchmark/index.js

```diff
@@ -0,0 +1,116 @@
+const path = require('path');
+const puppeteer = require('puppeteer');
+
+const { startServer, stopServer, PORT } = require('./staticServer');
+const { getBoundingClientRect, getCenterPositon, drag } = require('./puppeteerUtils');
+
+const getTracingResultsPath = name => path.resolve(__dirname, `./tracing-results/${name}.json`);
+
+const waitForSelectingMode = page => page.waitFor('.PatchWrapper-container.selecting');
+const pinSelector = (nodeId, pinName) => `#nodePinsOverlay_${nodeId} .PinOverlay[title=${pinName}]`;
+
+const width = 1300;
+const height = 850;
+(async () => {
+  await startServer();
+
+  const browser = await puppeteer.launch({
+    headless: false,
+    args: [`--window-size=${width},${height}`],
+  });
+
+  // setup
+  const page = await browser.newPage();
+
+  // @see https://reactjs.org/blog/2016/11/16/react-v15.4.0.html#profiling-components-with-chrome-timeline
+  // note that it will work only for non-production builds
+  await page.goto(`http://localhost:${PORT}/?react_perf`);
+  await page.setViewport({ width, height });
+  await page.waitForSelector('.Workarea');
+
+  // create a new patch
+  const patchName = 'my-test-patch';
+  await page.click('.ProjectBrowser button[title="Add patch"]');
+  await page.type('.PopupPrompt input[type=text]', patchName);
+  await page.click('.PopupPrompt button[type=submit]');
+
+  // open a patch group in project browser
+  const libname = 'xod/core';
+  await page.click(`.patch-group-trigger[title="${libname}"]`);
+  await page.waitFor(400); // wait for an animation to finish
+
+  const nodeToAdd = 'add';
+  // select a patch in project browser
+  await page.click(`.PatchGroupItem[title="${nodeToAdd}"] .PatchGroupItem__label`);
+  await page.waitFor('.PatchGroupItem.isSelected');
+
+  // placing nodes
+  const COLS = 14;
+  const ROWS = 5;
+  const nodeIds = [];
+
+  // TODO: specify categories for tarcing?
+  // @see https://github.com/GoogleChrome/puppeteer/blob/5e154dc835fcacdec6e680a695630352dae959dc/lib/Tracing.js#L36-L55
+  // TODO: User Timing API to marks placed here like this:
+  //    await page.evaluate(() => { window.performance.mark('added_node'); });
+  // show up in JSON, but are not visible when viewing profile in Chrome
+  await page.tracing.start({ path: getTracingResultsPath('adding_first_nodes') });
+  for (let col = COLS; col >= 0; col -= 1) {
+    if (col === 0) {
+      await page.tracing.start({ path: getTracingResultsPath('adding_last_nodes') });
+    }
+
+    for (let row = ROWS; row >= 0; row -= 1) {
+      // press 'add node' button
+      await page.click('.PatchGroupItem.isSelected .add-node');
+
+      await page.waitForSelector('.Node.is-selected');
+      const nodeHandle = await page.$('.Node.is-selected');
```

Comment:
Isn't better to name it `nodeHandle` -> `nodeElement`?

Reply:
Or if it returns not a real element, but a special wrapper or something like this, it could be named as "selector" / "cursor" and etc. IMHO
Maybe I'm just used to all that "handle" is about the processing of any actions, calls and other things. In short, a function.

Reply:
It's a [term from official docs](https://github.com/GoogleChrome/puppeteer/blob/master/docs/api.md#class-elementhandle).